src/platforms/common/stm32/serialno.c

```diff
@@ -64,3 +72,32 @@ void read_serial_number(void)
 #endif
 	serial_no[DFU_SERIAL_LENGTH - 1] = '\0';
 }
+
+static char *utoa_upper(unsigned value, char *str, int base)
+{
+	const char digits[] = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";
+
+	/* Check base is supported. */
+	if ((base < 2) || (base > 36)) {
+		str[0] = '\0';
+		return NULL;
+	}
+
+	/* Convert to string. Digits are in reverse order.  */
+	int l = 0;
+	do {
+		const unsigned remainder = value % base;
+		str[l++] = digits[remainder];
+		value = value / base;
+	} while (value != 0);
```

Comment:
This looks like it's begging to be a for loop:
```suggestion
	for (offset = 0; offset < 7U; ++offset) {
		const div_t result = div(value, base);
		const uint32_t remainder = result.rem;
		str[offset] = digits[remainder];
		if (!result.quot)
			break;
		value = result.quot;
	}
```

(This suggestion switches to using [div](https://en.cppreference.com/w/c/numeric/math/div) which should guarantee a fused divide+mod, though it does throw the cat among the pigeons on type correctness and safety because it's specified only for int/long/etc, so if you'd prefer to keep the explicit `/` and `%` to stay unsigned, that's fine with us)

Part of the reason for this switch is so we can enforce bounds on the buffer, improving the safety of this code.

Reply:
I could look at codegen to see the uldivmod call... it's a `udiv`+`mls` instruction pair. Maybe unsigned was used for indexing reasons, it's not like C has runtime rangechecking. If the number base conversion will run as fine with signed ints, then so be it.

Reply:
Yes, this works, and the `div()` function is `sdiv`+`mls` under the cover plus some stack manipulation.
```asm
div():
push	{r4}
sdiv	r4, r1, r2
mls	r1, r2, r4, r1
strd	r4, r1, [r0]
ldr.w	r4, [sp], #4
bx	lr
```
 I had to adjust for the offset being less by one now, but that's next on the menu. Tested live.